services/mlmodel/mlmodel.go

```diff
@@ -26,10 +30,127 @@ func init() {
 // the struct that will decode that map[string]interface{} correctly.
 type Service interface {
 	resource.Resource
-	Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error)
+	Infer(ctx context.Context, tensors ml.Tensors, input map[string]interface{}) (ml.Tensors, map[string]interface{}, error)
 	Metadata(ctx context.Context) (MLMetadata, error)
 }
 
+// TensorsToProto turns the ml.Tensors map into a protobuf message of FlatTensors.
+func TensorsToProto(ts ml.Tensors) (*servicepb.FlatTensors, error) {
+	pbts := &servicepb.FlatTensors{
+		Tensors: make(map[string]*servicepb.FlatTensor),
+	}
+	for name, t := range ts {
+		tp, err := tensorToProto(t)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert tensor to proto message")
+		}
+		pbts.Tensors[name] = tp
+	}
+	return pbts, nil
+}
+
+func tensorToProto(t *tensor.Dense) (*servicepb.FlatTensor, error) {
+	ftpb := &servicepb.FlatTensor{}
+	shape := t.Shape()
+	for _, s := range shape {
+		ftpb.Shape = append(ftpb.Shape, uint64(s))
+	}
+	// switch on data type of the underlying array
+	data := t.Data()
+	switch dataSlice := data.(type) {
+	case []int8:
+		unsafeByteSlice := *(*[]byte)(unsafe.Pointer(&dataSlice)) //nolint:gosec
+		data := &servicepb.FlatTensorDataInt8{}
+		data.Data = append(data.Data, unsafeByteSlice...)
+		ftpb.Tensor = &servicepb.FlatTensor_Int8Tensor{Int8Tensor: data}
```

Comment:
@acmorrow I think I end up doing a copy here anyway into the protobuf `FlatTensorDataInt8{}` slice using the append, but it doesn't seem like protobuf in Go has a way of "moving" the memory over to the new struct like C++?

Reply:
Even in C++ there is a copy this direction, into the protos, so I think you are doing what is expected. From what I can tell there is not a way to have this be zero copy.